pjlib/src/pj/string.c

```diff
@@ -143,6 +148,8 @@ PJ_DEF(char*) pj_strstr(const pj_str_t *str, const pj_str_t *substr)
 {
     const char *s, *ends;
 
+    PJ_ASSERT_RETURN(str->slen >= 0 && substr->slen >= 0, NULL);
+
     /* Special case when substr is zero */
     if (substr->slen == 0) {
```

Comment:
- change condition to `substr->slen <=0 || substr->slen > str->slen` ?
- also check for `str->slen <= 0`?

Reply:
ah, I keep forgetting that `PJ_ASSERT_RETURN` doesn't translate to `return` if `PJ_ENABLE_EXTRA_CHECK` is disabled.

ok, added check for `if (str->slen <= 0) return NULL;`.  There will be a minor behavior change though: `pj_strstr(empty string, empty string)` will now return NULL; whereas previously it will return `str->ptr`. But I feel the `NULL` return should be better since `str->ptr` may point to some arbitrary memory for empty string.

for the case `substr->slen > str->slen`, I don't think it's correct to return `str->ptr` since the substring clearly cannot exist in the string. The current behavior is already correct, to return `NULL`. If `substr->slen > str->slen`, `ends` will point to memory outside the string buffer, but it should be harmless since no memory access will be performed.